### PR TITLE
Polygon flashloan

### PR DESCRIPTION
**Pull request by KaymasJain**
TODO:
- [x] Test cases
- [x] Connector(Should be similar to [this](https://github.com/Instadapp/dsa-flashloan/blob/master/contracts/connectors/flashloan.sol))

### CODE REVIEW

File: contracts/receivers/aave-v2-receiver/main.sol [hunk 1]
```diff
@@ -126,6 +126,50 @@ contract AaveV2Migrator is MigrateResolver {
...
+        bytes calldata data
+    ) external isDSA {	
+        uint _length = _tokens.length;
+        require(_length == _amounts.length, "not-equal-length");
+        uint[] memory iniBal = new uint[](_length);
+        IERC20[] memory _tokenContracts = new IERC20[](_length);
+        for (uint i = 0; i < _length; i++) {
+            _tokenContracts[i] = IERC20(_tokens[i]);
+            iniBal[i] = _tokenContracts[i].balanceOf(address(this));
+            _tokenContracts[i].safeTransfer(msg.sender, _amounts[i]);
```
Comment (thrilok209) [inline contracts/receivers/aave-v2-receiver/main.sol:156]: What if we send `0xeeee...eee`? Should we convert it into wmatic address in the connector?

File: contracts/receivers/aave-v2-receiver/main.sol [hunk 2]
```diff
@@ -126,6 +126,50 @@ contract AaveV2Migrator is MigrateResolver {
...
+    modifier isDSA {
+        uint64 id = instaList.accountID(msg.sender);
+        require(id != 0, "not-dsa-id");
+        _;
+    }
+
+    function initiateFlashLoan(
+        address[] calldata _tokens,	
+        uint256[] calldata _amounts,	
+        uint _route, // no use of route but just to follow current flashloan pattern
```
Comment (thrilok209) [inline contracts/receivers/aave-v2-receiver/main.sol:146]:
```suggestion
        uint, // no use of route but just to follow current flashloan pattern
```

File: contracts/receivers/aave-v2-receiver/main.sol [hunk 3]
```diff
@@ -126,6 +126,50 @@ contract AaveV2Migrator is MigrateResolver {
...
+            _tokenContracts[i] = IERC20(_tokens[i]);
+            iniBal[i] = _tokenContracts[i].balanceOf(address(this));
+            _tokenContracts[i].safeTransfer(msg.sender, _amounts[i]);
+        }
+        CastData memory cd;
+        (cd.dsa, cd.route, cd.tokens, cd.amounts, cd.dsaTargets, cd.dsaData) = abi.decode(
+            data,
+            (address, uint256, address[], uint256[], address[], bytes[])
+        );
+        DSAInterface(msg.sender).cast(cd.dsaTargets, cd.dsaData, 0xB7fA44c2E964B6EB24893f7082Ecc08c8d0c0F87);
```
Comment (thrilok209) [inline contracts/receivers/aave-v2-receiver/main.sol:163]: Should we use castMigrate or cast? If we use cast, then we need to enable and disable migration address before and after the spells.

File: contracts/receivers/aave-v2-receiver/main.sol [hunk 4]
```diff
@@ -126,6 +126,50 @@ contract AaveV2Migrator is MigrateResolver {
...
+        IERC20[] memory _tokenContracts = new IERC20[](_length);
+        for (uint i = 0; i < _length; i++) {
+            _tokenContracts[i] = IERC20(_tokens[i]);
+            iniBal[i] = _tokenContracts[i].balanceOf(address(this));
+            _tokenContracts[i].safeTransfer(msg.sender, _amounts[i]);
+        }
+        CastData memory cd;
+        (cd.dsa, cd.route, cd.tokens, cd.amounts, cd.dsaTargets, cd.dsaData) = abi.decode(
+            data,
+            (address, uint256, address[], uint256[], address[], bytes[])
```
Comment (thrilok209) [inline contracts/receivers/aave-v2-receiver/main.sol:161]:
```suggestion
            (address, uint256, address[], uint256[], string[], bytes[])
```
Also, in DSAv2 the `cd.dsaTargets` is an `strings[]`. So, we need to change the data type in the decoding logic over here. Will handle the encoding part on the dsa-connect. If v2 then `strings[]`, If v1 then `address[]`

File: contracts/receivers/aave-v2-receiver/main.sol (hunk 3 above)
Comment (KaymasJain) [reply contracts/receivers/aave-v2-receiver/main.sol:163]: I was just thinking that we don't need to change the frontend code so trying to make things as simple as possible but what do you suggest? We can use castMigrate too or add the logic on cast itself for now?

File: contracts/receivers/aave-v2-receiver/main.sol (hunk 4 above)
Comment (KaymasJain) [reply contracts/receivers/aave-v2-receiver/main.sol:161]: Whoops! Updating. Have to also update the data we'll send from our end.

**Review by thrilok209 — APPROVED** [review]

**Pull request merged**